src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,372 @@
+                    if (transaction == null)
+                    {
+                        this.logger.LogDebug("Failure building specific consolidating transaction.");
+                        return null;
+                    }
+
+                    consolidationTransactions.Add(new ConsolidationTransaction
+                    {
+                        PartialTransaction = transaction,
+                        Status = ConsolidationTransactionStatus.Partial
+                    });
+
+                    roundNumber++;
+                    oneRound = unspentOutputs
+                        .Skip(roundNumber * FederatedPegSettings.MaxInputs)
+                        .Take(FederatedPegSettings.MaxInputs).ToList();
+                }
+
+                // Loop exits when we get a set of 50 that had a high enough amount, or when we run out of UTXOs aka a round less than 50
+
+                return consolidationTransactions;
+            }
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        private Transaction BuildConsolidatingTransaction(List<UnspentOutputReference> selectedInputs)
+        {
+            try
+            {
+                string walletPassword = this.walletManager.Secret.WalletPassword;
+                bool sign = (walletPassword ?? "") != "";
+
+                var multiSigContext = new TransactionBuildContext(new List<Recipient>())
+                {
+                    MinConfirmations = WithdrawalTransactionBuilder.MinConfirmations,
+                    Shuffle = false,
+                    IgnoreVerify = true,
+                    WalletPassword = walletPassword,
+                    Sign = sign,
+                    TransactionFee = FederatedPegSettings.ConsolidationFee,
+                    SelectedInputs = selectedInputs.Select(u => u.ToOutPoint()).ToList(),
+                    AllowOtherInputs = false,
+                    IsConsolidatingTransaction = true,
+                    Time = this.network.Consensus.IsProofOfStake 
+                        ? (uint?)selectedInputs.Max(x => x.Transaction.CreationTime).ToUnixTimeSeconds()
+                        : null
+                };
+
+                Transaction transaction = this.transactionHandler.BuildTransaction(multiSigContext);
+
+                this.logger.LogDebug("Consolidating transaction = {0}", transaction.ToString(this.network, RawFormat.BlockExplorer));
+
+                return transaction;
+            }
+            catch (Exception e)
+            {
+                this.logger.LogWarning("Exception when building consolidating transaction. Wallet state likely changed before calling: " + e);
+                return null;
+            }
+        }
+
+        /// <inheritdoc />
+        public void ProcessTransaction(Transaction transaction)
+        {
+            // TODO: It would be nice if there was a way to quickly check that the transaction coming in through here is FullySigned.
```

Comment:
As a first order approximation, you could count the number of signatures? That could trigger more expensive validation.